src/Microsoft.AspNetCore.SignalR.Core/HubEndPoint.cs

```diff
@@ -102,6 +105,9 @@ public class HubEndPoint<THub> where THub : Hub
 
         private async Task HubOnDisconnectedAsync(HubConnectionContext connection, Exception exception)
         {
+            // send close message before aborting the connection
+            await SendCloseAsync(connection, exception);
```

Comment:
Do we always send a close message?

Reply:
The spec of the close message is to always send a message on close. An error message tells the client if the connection closed from the server because of an error. Talk with @anurse if you don't like it.

Reply:
It's generally good form to send a close frame when a graceful disconnect occurs. It tells the client that the server intended to close the connection, rather than a mid-stream proxy killing it. For example, our client auto-reconnect logic should **not** attempt to reconnect if it sees a close frame with no error.

Reply:
Something about these changes to the HubEndPoint rub me the wrong way. I hate that we send the close message from so many different places. I really want this logic to be in a single place and I want various pieces of code to communicate via an exception. We have too many ways to send the close message right now.

Reply:
Oh wait, the handshake response is different from the close message...